url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/common/CircularDoubleLinkedList.java

```diff
@@ -0,0 +1,261 @@
+//===========================================================================
+//=-------------------------------------------------------------------------=
+//= Module history:                                                         =
+//= - January 3 2007 - Oscar Chavarro: Original base version                =
+//===========================================================================
+
+package vsdk.toolkit.common;
+
+class _CircularDoubleLinkedListNode<E> extends FundamentalEntity
+{
+    /// Check the general attribute description in superclass Entity.
+    public static final long serialVersionUID = 20070422L;
+
+    public E data;
+    public _CircularDoubleLinkedListNode<E> next;
+    public _CircularDoubleLinkedListNode<E> previous;
+}
+
+public class CircularDoubleLinkedList<E> extends FundamentalEntity
+{
+    /// Check the general attribute description in superclass Entity.
+    public static final long serialVersionUID = 20070422L;
+
+    private _CircularDoubleLinkedListNode<E> head;
+    private _CircularDoubleLinkedListNode<E> window;
+    private int lastAccessedIndex;
+    private int currentSize;
+
+    public CircularDoubleLinkedList()
+    {
+        head = null;
+        window = null;
+        currentSize = 0;
+        lastAccessedIndex = -1;
+    }
+
+    public int size()
+    {
+        return currentSize;
+    }
+
+    public void add(E e)
+    {
+        _CircularDoubleLinkedListNode<E> newContainer;
+        newContainer = new _CircularDoubleLinkedListNode<E>();
+        newContainer.data = e;
+        if ( head == null ) {
+            head = newContainer;
+            newContainer.next = newContainer;
+            newContainer.previous = newContainer;
+        }
+        else {
+            newContainer.previous = head.previous;
+            newContainer.next = head;
+            head.previous.next = newContainer;
+            head.previous = newContainer;
+        }
+        currentSize++;
+    }
+
+    public void insertBefore(E newElem, E pivot)
+    {
+        locateWindowAtElem(pivot);
+        _CircularDoubleLinkedListNode<E> newContainer;
+
+        newContainer = new _CircularDoubleLinkedListNode<E>();
+        newContainer.data = newElem;
+        lastAccessedIndex = -1;
+
+        if ( head == null ) {
+            head = newContainer;
+            newContainer.next = newContainer;
+            newContainer.previous = newContainer;
+        }
+        else if ( window == null || window == head ) {
+            window = head;
+            head = newContainer;
+            newContainer.previous = window.previous;
+            newContainer.next = window;
+            window.previous.next = newContainer;
+            window.previous = newContainer;
+        }
+        else {
+            newContainer.previous = window.previous;
+            newContainer.next = window;
+            window.previous.next = newContainer;
+            window.previous = newContainer;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 70:88, 216:234.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.